### PR TITLE
Connects Falkonry/tracker#815

### PR DESCRIPTION
**Pull request by dilpeshjain**

### CODE REVIEW

File: falkonryclient.egg-info/PKG-INFO [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Metadata-Version: 1.1
```
Comment (phagunbaya) [inline falkonryclient.egg-info/PKG-INFO]: @Dilpesh Don't commit *.egg-info files. Also add them in .gitignore file so that they are skipped.

File: falkonryclient/service/falkonry.py [hunk 2]
```diff
@@ -112,18 +112,20 @@ def add_input_data(self, datastream, data_type, options, data):
...
-        if 'hasMoreData' in options and options['hasMoreData'] is True:
-            hasMoreData = 'true'
-        else:
-            hasMoreData = 'false'
-
-
-        url = '/Datastream/' + str(datastream) + '?streaming=' + streaming + '&hasMoreData=' + hasMoreData
+        url = self.get_add_input_data_url(datastream, options)
+
+        # if 'streaming' in options and options['streaming'] is True:
```
Comment (phagunbaya) [inline falkonryclient/service/falkonry.py]: Remove commented part

File: test/dataNarrow.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{"time" :"2016-03-01 01:01:01", "signal" : "current", "value" : 12.4, "car" : "car1"}
+{"time" :"2016-03-01 01:01:01", "signal" : "vibration", "value" : 3.4, "car" : "car1"}
+{"time" :"2016-03-01 01:01:01", "signal" : "state", "value" : on, "car" : "car1"}
```
Comment (phagunbaya) [inline test/dataNarrow.json]: `on` and `off` should be in double quotes.

File: test/TestDatastream.py [hunk 4]
```diff
@@ -467,28 +478,28 @@ def test_delete_datastream_by_id(self):
 
     # Create Datastream microseconds precision
     def test_create_datastream_narrow_style_multiple_entity(self):
-        fclient = FClient(host=host, token=token,options=None)
+        fclient = FClient(host=host, token=token, options=None)
         datastream = Schemas.Datastream()
+        datastream.set_name('Motor Health' + str(random.random()))
+        datastream.set_time_precision('micro')  # set 'micro' for microseconds precision
```
Comment (phagunbaya) [inline test/TestDatastream.py]: If no one is setting it then default will be `micro` and will be set by merlin right ?

File: test/TestDatastream.py [hunk 5]
```diff
@@ -383,12 +371,34 @@ def test_get_datastream_by_id(self):
...
+        signal.set_valueIdentifier("value")
         field.set_signal(signal)
         datasource.set_type("STANDALONE")
         field.set_time(time)
         datastream.set_datasource(datasource)
         datastream.set_field(field)
 
+
+
+        # fclient = FClient(host=host, token=token,options=None)
```
Comment (phagunbaya) [inline test/TestDatastream.py]: Remove commented code

File: README.md [hunk 6]
```diff
@@ -108,23 +108,23 @@ datastream.set_field(field)
 createdDatastream = fclient.create_datastream(datastream)
 ```
 
-#### Create Datastream for narrow/historian style data from multiple things
+#### Create Datastream for narrow/historian style data from multiple entitys
```
Comment (aniket-amrutkar) [inline README.md]: typo it should be "entities"

File: README.md [hunk 7]
```diff
@@ -505,24 +524,33 @@ falkonry   = Falkonry('http://localhost:8080', 'auth-token')
 datastreamId = 'id of the datastream'
 
 #add data to Datastream
-stream   = io.open('./data.json')
+stream   = io.open('./data.csv')
```
Comment (aniket-amrutkar) [inline README.md]: this should be .json

File: README.md [hunk 8]
```diff
@@ -26,10 +26,10 @@ $ pip install falkonryclient
     * Delete Datastream
     * Add EntityMeta to a Datastream
     * Get EntityMeta of a Datastream
-    * Add historical input data (json format) to Datastream (Used for model revision) 
-    * Add historical input data (csv format) to Datastream (Used for model revision) 
-    * Add historical input data (json format) from a stream to Datastream (Used for model revision) 
-    * Add historical input data (csv format) from a stream to Datastream (Used for model revision) 
+    * Add narrow input data (json format) to multi entity Datastream
```
Comment (aniket-amrutkar) [inline README.md]: Keep titles as it is, as they are more descriptive 
Add historical input data (json format) to Datastream (Used for model revision)

File: README.md [hunk 9]
```diff
@@ -505,24 +524,33 @@ falkonry   = Falkonry('http://localhost:8080', 'auth-token')
...
 # set hasMoreData to True if data is sent in batches. When the last batch is getting sent then set  'hasMoreData' to False. For single batch upload it shpuld always be set to False
-options = {'streaming': False, 'hasMoreData':False}   
-inputResponse = falkonry.add_input_data(datastreamId, 'json', options, stream)
+
+options = {'streaming': False,
+           'hasMoreData': False,
+           'timeFormat': "YYYY-MM-DD HH:mm:ss",
+           'timeZone': "GMT",
+           'timeIdentifier': "time"}
+inputResponse = falkonry.add_input_data(datastreamId, 'csv', options, stream)
```
Comment (aniket-amrutkar) [inline README.md]: it should be
inputResponse = falkonry.add_input_data(datastreamId, 'json', options, stream)

File: README.md [hunk 10]
```diff
@@ -546,10 +579,10 @@ inputResponse = falkonry.add_input_data(datastreamId, 'csv', options, stream)
 Data :
 
 ```
-    {"time" :"2016-03-01 01:01:01", "tag" : "signal1_thing1", "value" : 3.4}
-    {"time" :"2016-03-01 01:01:01", "tag" : "signal2_thing1", "value" : 1.4}
-    {"time" :"2016-03-01 01:01:02", "tag" : "signal1_thing2", "value" : 9.3}
-    {"time" :"2016-03-01 01:01:02", "tag" : "signal2_thing2", "value" : 4.3}
+    {"time" :"2016-03-01 01:01:01", "tag" : "signal1_entity1", "value" : 3.4}
```
Comment (aniket-amrutkar) [inline README.md]: Change formats in live cases also.
Why did you change it only in historical ingestion?

File: test/TestAddData.py [hunk 11]
```diff
@@ -2,16 +2,16 @@
...
 
 
 class TestAddData(unittest.TestCase):
 
     def setUp(self):
         pass
 
-    # Add historical input data (json format) to Datastream (Used for model revision)
-    def test_add_data_historical_json(self):
+    # Add narrow input data (json format) to multi entity Datastream
```
Comment (aniket-amrutkar) [inline test/TestAddData.py]: Please keep previous comments as they are more descriptive.

File: test/TestAddData.py [hunk 12]
```diff
@@ -24,17 +24,28 @@ def test_add_data_historical_json(self):
...
+        field.set_entityIdentifier('car')
         datastream.set_datasource(datasource)
         datastream.set_field(field)
 
         try:
             datastreamResponse = fclient.create_datastream(datastream)
             try:
-                data = '{"time" :"2016-03-01 01:01:01", "current" : 12.4, "vibration" : 3.4, "state" : "On"}'
-                options = {'streaming': False, 'hasMoreData':False}
+                # data = "time,signal,car,value " + "\n" + "2016-03-01 01:01:01,signal1,car1,3.4" + "\n" + "2016-03-01 01:01:01,signal2,car1,1.4" + "\n" + "2016-03-01 01:01:01,signal1,car2,1.4" + "\n" + "2016-03-01 01:01:01,signal2,car2,1.4"
```
Comment (aniket-amrutkar) [inline test/TestAddData.py]: remove commented code

File: test/TestAddData.py [hunk 13]
```diff
@@ -24,17 +24,28 @@ def test_add_data_historical_json(self):
...
+                data = '{"time" : "2016-03-01 01:01:01", "signal" : "current", "value" : 12.4, "car" : "unit1"}'
+                options = {'streaming': False,
+                           'hasMoreData': False,
+                           'timeFormat': time.get_format(),
```
Comment (aniket-amrutkar) [inline test/TestAddData.py]: Use different time format than datastream's time format in one of the test case

File: test/TestAddData.py [hunk 14]
```diff
@@ -90,14 +196,147 @@ def test_add_data_historical_csv(self):
...
+            try:
+                data = "time,current,vibrarion,state,car " + "\n" + "2016-03-01 01:01:01,12.4,3.4,on,car1" + "\n" + "2016-03-01 01:01:01,31.2,1.4.off,car1" + "\n" + "2016-03-01 01:01:01,24,3.2,on,car2" + "\n" + "2016-03-01 01:01:01,31,3.4.off,car2"
+                options = {'streaming': False,
+                           'hasMoreData': False,
+                           'timeFormat': time.get_format(),
+                           'timeZone': time.get_zone(),
+                           # 'timeIdentifier': time.get_identifier(),
+                           'entityIdentifier': 'car'}
+                response = fclient.add_input_data(datastreamResponse.get_id(), 'csv', options, data)
+
```
Comment (aniket-amrutkar) [inline test/TestAddData.py]: you should add assertion on this line too

File: test/TestAddData.py [hunk 15]
```diff
@@ -90,14 +196,147 @@ def test_add_data_historical_csv(self):
...
+            try:
+                data = "time,current,vibrarion,state,car " + "\n" + "2016-03-01 01:01:01,12.4,3.4,on,car1" + "\n" + "2016-03-01 01:01:01,31.2,1.4.off,car1" + "\n" + "2016-03-01 01:01:01,24,3.2,on,car2" + "\n" + "2016-03-01 01:01:01,31,3.4.off,car2"
+                options = {'streaming': False,
+                           'hasMoreData': False,
+                           'timeFormat': time.get_format(),
+                           # 'timeZone': time.get_zone(),
+                           'timeIdentifier': time.get_identifier(),
+                           'entityIdentifier': 'car'}
+                response = fclient.add_input_data(datastreamResponse.get_id(), 'csv', options, data)
+
```
Comment (aniket-amrutkar) [inline test/TestAddData.py]: same here
Add assertion on this line too

Comment (aniket-amrutkar) [reply test/TestAddData.py]: Add assertions at all negative test cases

File: test/TestAddDataStream.py [hunk 16]
```diff
@@ -3,14 +3,15 @@
...
+token = 'npp766l2hghmhrc7ygrbldjnkb9rn7mg'                       # auth token
+
 
 class TestAddDataStream(unittest.TestCase):
 
     def setUp(self):
         pass
 
-    # Add historical input data (json format) from a stream to Datastream (Used for model revision)
+    # Add historical wide input data (json format) to single entity Datastream (Used for model revision)
```
Comment (aniket-amrutkar) [inline test/TestAddDataStream.py]: add proper comment

File: test/TestAddDataStream.py [hunk 17]
```diff
@@ -73,7 +79,62 @@ def test_add_historical_csv_data_stream(self):
...
+        field.set_signal(signal)
+        datasource.set_type("STANDALONE")
+        field.set_time(time)
+        field.set_entityIdentifier('car')
+        datastream.set_datasource(datasource)
+        datastream.set_field(field)
+        try:
+            datastreamResponse = fclient.create_datastream(datastream)
+            try:
+                data = io.open('./dataMultientity.csv')
```
Comment (aniket-amrutkar) [inline test/TestAddDataStream.py]: There is no file called dataMultientity.csv in the commit.
you have replaced all "thing" with "entity" which caused this bug.

File: test/TestAssessment.py [hunk 18]
```diff
@@ -289,4 +289,7 @@ def test_delete_assessment(self):
     else:
         from ..falkonryclient import schemas as Schemas
         from ..falkonryclient import client as FClient
-    unittest.main()
\ No newline at end of file
+    unittest.main()
+else:
```
Comment (aniket-amrutkar) [inline test/TestAssessment.py]: Why do we want this code?

File: test/TestDatastream.py [hunk 19]
```diff
@@ -70,10 +70,7 @@ def test_create_datastream_narrow_style_single_entity(self):
         time.set_identifier("time")                                 # set time identifier of the datastream
         time.set_format("iso_8601")                                 # set time format of the datastream
         field.set_time(time)
-        signal.set_delimiter(None)                                  # set delimiter to None
-        signal.set_tagIdentifier("tag")                             # set tag identifier
-        signal.set_valueIdentifier("value")                         # set value identifier
-        signal.set_isSignalPrefix(False)                            # as this is single entity, set signal prefix flag to false
+        # signal.set_valueIdentifier("value")                         # set value identifier
```
Comment (aniket-amrutkar) [inline test/TestDatastream.py]: Why this is commented ?
For narrow format valueIdentifier and signalIdentifier both are compulsory

File: test/TestDatastream.py [hunk 20]
```diff
@@ -123,10 +117,7 @@ def test_create_datastream_narrow_style_multiple_entity(self):
         time.set_identifier("time")                                 # set time identifier of the datastream
         time.set_format("iso_8601")                                 # set time format of the datastream
         field.set_time(time)
-        signal.set_delimiter("_")                                  # set delimiter to None
-        signal.set_tagIdentifier("tag")                             # set tag identifier
         signal.set_valueIdentifier("value")                         # set value identifier
```
Comment (aniket-amrutkar) [inline test/TestDatastream.py]: Same here.
How this test case was passed ? 
If you are providing valueIdentifer then signalIdentifier is compulsory

File: test/TestDatastream.py (hunk 4 above)
Comment (dilpeshjain) [reply test/TestDatastream.py]: By default is it millis and yes it will be set by merlin.

File: test/TestDatastream.py [hunk 21]
```diff
@@ -500,10 +511,7 @@ def test_create_datastream_narrow_style_multiple_entity(self):
             self.assertEqual(fieldResponse.get_entityIdentifier(),"entity",'Invalid entity identifier object after creation')
             self.assertEqual(fieldResponse.get_entityName(),None,'Invalid entity name object after creation')
             signalResponse = fieldResponse.get_signal()
-            self.assertEqual(signalResponse.get_delimiter(),signal.get_delimiter(), 'Invalid delimiter after object creation')
-            self.assertEqual(signalResponse.get_tagIdentifier(),signal.get_tagIdentifier(), 'Invalid tag identifier after object creation')
             self.assertEqual(signalResponse.get_valueIdentifier(),signal.get_valueIdentifier(), 'Invalid value identifier after object creation')
```
Comment (aniket-amrutkar) [inline test/TestDatastream.py]: Also add assertion for signalIdentifier

File: test/TestDatastream.py (hunk 20 above)
Comment (phagunbaya) [reply test/TestDatastream.py]: @dilpeshjain Did you verify this ?
